modules/javafx.base/src/test/java/test/util/memory/JMemoryBuddy.java

```diff
@@ -0,0 +1,317 @@
+
+        if (folder1.exists()) return folder1.getAbsolutePath();
+        if (folder2.exists()) return folder2.getAbsolutePath();
+        return ".";
+    }
+
+    static void createGarbage() {
+        LinkedList list = new LinkedList<Integer>();
+        int counter = 0;
+        while (counter < garbageAmount) {
+            counter += 1;
+            list.add(1);
+        }
+    }
+
+    /**
+     * Checks whether the content of the WeakReference can be collected.
+     * @param weakReference The WeakReference to check.
+     */
+    public static void assertCollectable(WeakReference weakReference) {
+        if (!checkCollectable(weakReference)) {
+            AssertCollectable assertCollectable = new AssertCollectable(weakReference);
+            createHeapDump();
+            throw new AssertionError("Content of WeakReference was not collected. content: " + weakReference.get());
+        }
+    }
+
+    /**
+     * Checks whether the content of the WeakReference can be collected.
+     * @param weakReference The WeakReference to check.
+     * @return Returns true, when the provided WeakReference can be collected.
+     */
+    public static boolean checkCollectable(WeakReference weakReference) {
+        return checkCollectable(steps, weakReference) > 0;
+    }
+
+    private static int checkCollectable(int stepsLeft, WeakReference weakReference) {
+        int counter = stepsLeft;
+
+        if (weakReference.get() != null) {
+            createGarbage();
+            System.gc();
+            System.runFinalization();
+        }
+
+        while (counter > 0 && weakReference.get() != null) {
+            try {
+                Thread.sleep(sleepDuration);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            counter = counter - 1;
+            createGarbage();
+            System.gc();
+            System.runFinalization();
+        }
+
+        if (weakReference.get() == null && counter < steps / 3) {
+            int percentageUsed = (int) ((steps - counter) / steps * 100);
+            System.out.println("Warning test seems to be unstable. time used: " + percentageUsed + "%");
+        }
+
+        return counter;
+    }
+
+    /**
+     * Checks whether the content of the WeakReference can not be collected.
+     * @param weakReference The WeakReference to check.
+     */
+    public static void assertNotCollectable(WeakReference weakReference) {
+        if (!checkNotCollectable(weakReference)) {
+            throw new AssertionError("Content of WeakReference was collected!");
+        }
+    }
+
+    /**
+     * Checks whether the content of the WeakReference can not be collected.
+     * @param weakReference The WeakReference to check.
+     * @return Returns true, when the provided WeakReference can not be collected.
+     */
+    public static boolean checkNotCollectable(WeakReference weakReference) {
+        createGarbage();
+        System.gc();
```

Comment:
I've added runFinalization followed with another System.gc (otherwise it wouldn't make any difference)
I'm quite sure it's avoiding an error in the library, but I don't have a unit-test for it yet.

It's a tradeoff between speed and reliability. The library tries to avoid false-negative test results at all costs.
false-positive results are tolerated for speed if they are "sufficiently reliable".
The Library has tests for assertNotCollectable, which are executed with every commit for different JVM's, and I don't remember any false-positive results (which would be visible as a negative test result, because the unit-test checks for failing tests). For that reason, we can assume they are "sufficiently reliable" despite the theoretical problem.

Reply:
That sounds fine then.

As soon as you've pushed the follow-on fix (to add the `runFinalization` and a second `gc`) I'll finish my review.

Reply:
I forgot to push my changes yesterday. They are pushed now!